source/features/useful-not-found-page.tsx

```diff
@@ -107,19 +123,55 @@ function init(): false | void {
 	for (let i = bar.children.length - 2; i >= 0; i--) {
 		void checkAnchor(bar.children[i] as HTMLAnchorElement);
 	}
+}
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+async function showHelpfulLinks(): Promise<void> {
+	// What we should try to do to be helpful:
+	//  1. Check if the file is on the default branch
+	//  2. Check the current branch's history for the file (was it deleted or renamed)
+
+	const urlToFileOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlToFileOnDefaultBranch) {
+		select('h2')!.after(
+			<p className="container mt-4 text-center">
+				<a href={urlToFileOnDefaultBranch}>This {getType()}</a> exists on the default branch.
+			</p>,
+		);
+		return;
 	}
 
-	if (['tree', 'blob', 'edit'].includes(parts[2])) {
-		// File might not be available on the current branch
-		// GitHub already redirects /tree/ and /blob/ natively
-		void addDefaultBranchLink(bar);
+	const change = await getLatestChangeToFile();
+	if (change) {
+		const {fileInfo, commitInfo} = change;
+		const url = new GitHubURL(location.href);
+
+		url.assign({route: 'commits'});
+		const commitHistory = <a href={url.toString()}>Commit history</a>;
+		url.assign({route: 'blob', branch: commitInfo.parents[0].sha, filePath: url.filePath});
+		const lastVersion = <a href={fileInfo.status === 'removed' ? fileInfo.blob_url : url.toString()}>This {getType()}</a>;
```

Comment:
```suggestion
		const lastVersionUrl = fileInfo.status === 'removed' ? fileInfo.blob_url : url.toString();
		const lastVersion = <a href={lastVersionUrl}>The {getType()}</a>;
```

I generally don't like single-use variables but here it's better than having to wrap the JSX (which is a little long)